Returning MSet instead of sorted Docids after re-ranking (Ticket #734)

AyushP123 opened the pull request:
Changes made to MSet: Added two functions to the public API set_new_weights and re_rank to MSet along with a helper function set_new_weights_helper.
1. set_new_weights: assigns new weights to items
2. re_rank sorts the MSet by sorting items on the basis of their weights.

Changes made to Letor: 
1. Keeping the vector<FeatureVector> in the original order. Ranker only updates weights.
2. New weights are obtained from vector<FeatureVector> and assigned to items and it is sorted according to the wieghts subsequently.
3. Printing the docids from MSet to show the updated order after ranker re-ranks MSet

jaylett commented on xapian-core/include/xapian/mset.h:
```diff
@@ -47,6 +47,10 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
     // Helper function for fetch() methods.
     void fetch_(Xapian::doccount first, Xapian::doccount last) const;
 
+  private:
```
Isn't this already `private`? There's nothing to change from the default for class members.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +78,31 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+    }
+
+    /**
+     * Sorts the MSet::Internal::items according to their weights thereby sorting MSet
```
Documentation for functions in the public API shouldn't generally mention the wrapped internal class. Just say something like "Sort the items in this MSet by weight. Use after calling set_new_weights()."

AyushP123 replied:
`/** Sorts the list of documents in MSet according to their weights.
     *
     *  Use after calling MSet::replace_weights.
     `*/``
will be using this string if it is acceptable to you

jaylett commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +78,31 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+     * It updates weights of all the elements in items as well as updating the max_attained
+     */
+    template <typename Iterator>
+    void set_new_weights(Iterator first, Iterator last) {
```
I think `replace_weights()` reads slightly better, although others may disagree.

AyushP123 replied:
:+1:

jaylett commented on xapian-letor/bin/xapian-rank.cc:
```diff
@@ -207,13 +207,16 @@ try {
+    for (Xapian::MSetIterator m = mset.begin();m != mset.end();++m){
 
+	Xapian::docid did = *m;
+	cout << "Rank " << i << ": " << did << endl;
```
While we're here, it would probably be useful (in a separate commit) to update this so it matches the output of `quest` (see `xapian-core/examples/quest.cc`).

AyushP123 replied:
The format given in quest.cc is "docid doc_weight \n doc_data". Should I follow just "docid doc_weight" ??

AyushP123 replied:
:+1:

jaylett commented on xapian-letor/ranker/listnet_ranker.cc:
```diff
@@ -193,6 +193,6 @@ ListNETRanker::rank_fvv(const std::vector<FeatureVector> & fvv) const {
 	testfvv[i].set_score(listnet_score);
     }
-    std::sort(testfvv.begin(), testfvv.end(), &Ranker::scorecomparer);
+    //std::sort(testfvv.begin(), testfvv.end(), &Ranker::scorecomparer);
```
Just delete this line; you don't need to keep commented code, because we can find it from the git history if we need.

AyushP123 replied:
:+1:

jaylett commented on xapian-letor/ranker/ranker.cc:
```diff
@@ -351,19 +351,20 @@ Ranker::labelcomparer(const FeatureVector & firstfv, const FeatureVector& second
     std::vector<FeatureVector> rankedfvv = rank_fvv(fvv);
 
-    std::vector<Xapian::docid> rankeddid;
+    std::vector<double> weights;
```
This still creates a vector of doubles from a vector of things that contain those doubles. It should be possible to create an iterator over the vector which returns the letor score on `*`-deference, so it can be passed in directly to `set_new_weights()`. That said, it's getting into STL use that is quite subtle and definitely beyond what I've done in the past.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,23 @@ MSet::Internal::read_docs() const
+void
+MSet::Internal::item_set_weight(Xapian::Internal::MSetItem& item_,double wt_)
+{
+    item_.wt = wt_;
```
This doesn't adjust `max_possible` or `max_attained`, if that's needed.

AyushP123 replied:
:+1:

AyushP123 commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,23 @@ MSet::Internal::read_docs() const
+}
+
+void
+MSet::Internal::re_rank()
```
sort_by_weights() should be a better name??

jaylett replied:
Having consistency across the API is better, so `sort_by_relevance()` would be the right choice here since we already have a similar-named function working in the same way.

AyushP123 replied:
:+1:  .... renamed it to sort_by_relevance()

AyushP123 commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,23 @@ MSet::Internal::read_docs() const
 }
 
+void
+MSet::Internal::set_new_weights(double wt, int i) {
```
I am not able to set the weight of the particular item as `item[i].wt = wt`that is why I am using item_set_weight()

jaylett replied:
You should be able to set it directly from within any `MSet::Internal` method. However, if you really cannot, you should overload `MSet::Internal::set_item_weight()` rather than call one `set_item_weight_()`.

AyushP123 replied:
:+1:  .... have eliminated set_item_weight() helper

jaylett commented on xapian-core/include/xapian/mset.h:
```diff
@@ -83,25 +82,24 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
 	{
-	    set_new_weights_helper(*it, i);
-	    ++i;
+	    set_item_weight(i, (*it).get_score());
```
This isn't a great idea, because then the only iterators that make sense are ones that have `.get_score()`. You should be able to create a custom iterator inside `ranker/ranker.cc` which dereferences to `double` (in other words, which has a `double operator*()` method). Then the public API for changing weights can be used by anyone with a `vector<double>`, for instance, which is more what I'd expect.

AyushP123 replied:
:+1:

jaylett commented:
@AyushP123 a couple of minor points, plus the more subtle one about `max_attained` (which should become more obvious when writing tests). The other, about the iterator dereferencing to a `double`, should also be more obvious when you come to write tests for the `MSet` changes, so I'd definitely focus on unit tests next.

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +77,30 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
 
+    /** Assigns new weights to items
+     * This functon is used in letor to update the weights assigned by the ranker
+     * It updates weights of all the elements in items as well as updating the max_attained
```
You're writing a description of the public API here that's supposed to be useful for people trying to use the API.  Discussing that a method is used used by one particular piece of software using the public API isn't likely to be helpful here.

Also `items` is a purely internals concept.  Dropping "in items" would be better.

This reads oddly: "as well as updating the max_attained" - I'd either drop "the" or say something like "the max_attained value".

There's no documentation of what the parameters are meant to be, and being a template there's not even type information to give a clue.  All the user reading the docs can tell is that they're iterators but there's no clear indication of what sort of iterators or what over.

Also a sentence should end in a full stop.

AyushP123 replied:
` /** Assigns new weights and updates MSet.
     *
     *  Dereferencing the Iterator should return a double.
     *
     *  Iterator should dereference weights in the original order of documents.
     *  Weights are assigned as per the original order of documents.
     *
     *  @param begin	Begin iterator.
     *  @param end	End iterator.
     *
     *  @exception Xapian::InvalidArgument is thrown if the total number of
     *		    elements in the input doesn't match the total number of
     *		    documents in MSet.
     */`
will be using this docstring if it is acceptable to you

ojwb replied:
That would be more readable as a comment if you put a line of 3 backticks ( ```) before and after as that renders as a preformatted block - I did an edit to do that.

The `Iterator ...` paragraph seems a bit woolly - weights don't get dereferenced (they're returned by the iterator when dereferenced), the "original order of the documents" could be taken to mean "docid order" rather than the order in the `MSet`, and it talks about the order twice which might make the reader think it's talking about two different things.

I'd suggest `The weights returned by the iterator are assigned to elements of the MSet in rank order.`

You've also indented the continuation of the `@exception` line too much - `elements` and `documents` should line up under `Xapian`.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +77,30 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+     */
+    template <typename Iterator>
+    void replace_weights(Iterator first, Iterator last) {
+	if (std::distance(first, last) != size())
```
This is fine for a random access iterator, but for a input iterator it requires checking each element to find how many there are which doesn't seem ideal.  In particular it means the iterator needs to be copyable.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +77,30 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+    template <typename Iterator>
+    void replace_weights(Iterator first, Iterator last) {
+	if (std::distance(first, last) != size())
+	{
```
Xapian coding standard is that `{` should be on the same line as `if`, etc.  Also `for` a few lines below.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +77,30 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+	if (std::distance(first, last) != size())
+	{
+	    throw std::invalid_argument("Number of weights assigned don't match the number of items.Aborting");
+	    return;
```
`return` is dead here.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +77,30 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+    void replace_weights(Iterator first, Iterator last) {
+	if (std::distance(first, last) != size())
+	{
+	    throw std::invalid_argument("Number of weights assigned don't match the number of items.Aborting");
```
Should be `Xapian::InvalidArgumentError`.  Also "Aborting" shouldn't be in the message (the thrower can't tell how the exception will be handled).  And the verb is wrong - "don't" should be "doesn't" since the subject is "number of weights" which is singular.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +77,30 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+	    throw std::invalid_argument("Number of weights assigned don't match the number of items.Aborting");
+	    return;
+	}
+	int i = 0;
```
`Xapian::doccount` would be better than `int`.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +77,30 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+	    return;
+	}
+	int i = 0;
+	for (Iterator it = first;it != last;++it,++i)
```
This requires the iterator to be copyable.  I'd suggest using a `while` loop and just using `first` as the iterator you advance.

Also whitespace issues.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +77,30 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+     * Sorts the items in MSet according to their weights. Use afer calling set_new_weights();
+     */
+
+    void sort_by_relevance();
```
There shouldn't be a blank line between the comment and the function.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +77,30 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+	int i = 0;
+	for (Iterator it = first;it != last;++it,++i)
+	{
+	    set_item_weight(i, (*it).get_score());
```
This requires that the iterator is over objects supporting a `get_score()` method, which seems unnecessarily restrictive.

AyushP123 replied:
:+1:

ojwb commented on xapian-letor/bin/xapian-rank.cc:
```diff
@@ -207,13 +207,16 @@ try {
-    for (auto&& did : ranked_docids)
-	cout << "Rank " << ++rank << ": " << did << endl;
+    int i = 1;
+    for (Xapian::MSetIterator m = mset.begin();m != mset.end();++m){
```
Whitespace again.

AyushP123 replied:
:+1:

ojwb commented on xapian-letor/bin/xapian-rank.cc:
```diff
@@ -207,13 +207,16 @@ try {
-    rank = 0;
-    for (auto&& did : ranked_docids)
-	cout << "Rank " << ++rank << ": " << did << endl;
+    int i = 1;
```
I'd use `MSetIterator::get_rank()` rather than counting the rank yourself.

AyushP123 replied:
:+1:  ..... Currently we are using examples/quest.cc format .... we are not printing ranks

ojwb commented on xapian-letor/ranker/ranker.cc:
```diff
@@ -351,19 +351,15 @@ Ranker::labelcomparer(const FeatureVector & firstfv, const FeatureVector& second
-	rankeddid.push_back(rankedfvv[i].get_did());
-    }
-    return rankeddid;
+    mset.replace_weights(rankedfvv.begin(), rankedfvv.end());
```
As noted above, this requires that anyone using the API to implement iteration over objects providing a `get_score()` method.  But avoiding the temporary vector just to hold the weights is good.

Better would be for the new xapian-core API to just dereference the iterator to get the weight, and have a "shim" iterator class here which does the dereferencing.  Something like:

```
class ScoreIterator {
    std::vector<FeatureVector>::const_iterator it;

  public:
    explicit ScoreIterator(const std::vector<FeatureVector>::const_iterator it_) : it(it_) { }
 
    double operator*() const { return it->get_score(); }

    ScoreIterator& operator++() { ++it; return this; }

    bool operator==(const ScoreIterator& o) { return it == o.it; }

    // forward other methods to "it"...
};
```

Although it may look like this involves extra code and extra function calls, because it's all inlined via the template, the extra stuff just optimises away.

Being able to write simple shims like the above is the reason why it's good to avoid unnecessary extra requirements on the iterators.

AyushP123 replied:
:+1:

AyushP123 commented:
I have incorporated all the changes suggested .... if there are no further issues then we should merge this PR

jaylett commented:
@AyushP123 you still haven't written tests for the changes to MSet.

AyushP123 commented:
We should merge this PR now so that we could shift focus to the other PR regarding ticket #733

jaylett commented:
@AyushP123 please don't keep asserting that things must be merged quickly. It's moving too quickly that's causing this to take so long; instead of racing to get this merged, you need to take sufficient time to ensure that things are not just done, but done well.

Things I would still like to see improved:

 * the comments in `MSet::Internal::set_item_weight()` simply describe the code. They need to explain _why_ this is the right approach. In the calculation of `max_attained`, there is an assumption that this function is only called from `replace_weights()`, which will replace all the weights and hence correctly calculate the stat. This must be stated clearly. (The comment for `max_possible` literally just states the code it comments. Again, a comment as to why this is appropriate is needed.)

 * the doc string for `MSet::replace_weights()` could I think be more descriptive. I think it should explicitly state that the iterator should be over the MSet items (ie documents) in the order they are in the MSet, so people don't have to scratch their heads and assume that's how it will work. (It's the only logical way it would, be being explicit is important in helping people get to grips with something like this.)

 * If you can in your editor, spell check the comments and doc strings. I've spotted ["afer" (shoudl be "after")](https://github.com/xapian/xapian/pull/156/files#diff-8e425dcb34550c8a50d1f1f9e6cb113cR105), and I'm sure there are more.

 * Test that `max_possible` is updated when it should be.

 * Commits that match the structure of the change. I think there are two commits here: adding the new `MSet` functionality (including tests), and changing the Letor system to use it. This is actually very easy to rebuild because the changes for each commit are in a different part of the source tree, so we should do that once you've addressed the other concerns, plus anything that @ojwb wants to add.

Finally, for the tests I meant to get rid of the comments as well as being more terse with constructing the vectors. For instance:

```c++
    // open the database (in this case a simple text file
    // we prepared earlier)

    Xapian::Database mydb(get_database("apitest_onedoc"));
```

Every other test just opens the database, without commenting (in fact, they don't even store the `Database` object directly — they just make an `Enquire` straight out of it). We should assume that common Xapian operations are readable by people looking at the tests, I think. (Otherwise the tests all get verbose.) None of the comments in the new tests add anything to the code; if you want to comment why `max_attained` should hit the value you're testing for, then that would be appropriate, but just saying that it should have been updated correctly is what the code already indicates.

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,77 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
 DEFINE_TESTCASE(replace_weights, backend) {
-    // open the database (in this case a simple text file
-    // we prepared earlier)
 
```
You can get rid of these blank lines, too. Try to make it look as much like `sortfunctortemote1` above as possible.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,77 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
     enquire.set_query(Xapian::Query("word"));
 
-    // retrieve the top ten results (we only expect one)
+    // retrieve the top ten results (we only expect one).
```
Is this comment needed? There's only one document in the database `apitest_onedoc`, so even that detail should be fairly clear.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,77 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
 
-    // assigning new weight of -1 to the document.
+    /* assigning new weight of -1 to the document.
+     * a negative weight was chose to avoid matching the positive weight returned by enquire.
```
"Matching" is misleading here. Also, I'm slightly uncomfortable with using a negative "weight" in our example, given that weights are never negative elsewhere in Xapian. We might be better off choosing a replacement weight that is explicitly lower than the non-letor weight.

AyushP123 replied:
would 0 be a good example weight??

jaylett replied:
No, I'd use a strictly positive weight. You might want to figure out what the actual weight of the single MSet item is in this case and just take a little off that. (It won't change unless we change the default weighting scheme, which is incredibly unlikely.)

ojwb replied:
Yes, let's avoid negative weights.  Negative weights should perhaps be rejected if you try to set them, though that raises issues with the `MSet` weights being partially updated when we discover a negative one.

If you want a weight that's definitely lower than the maximum attained weight, I'd suggest just halving an existing weight.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,77 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
     mymset.replace_weights(weights.begin(), weights.end());
     Xapian::MSetIterator i = mymset.begin();
-    // checking wether the weight was assigned correctly
+    // The weight for the document should be updated to double(-1).
```
This comment says nothing that the following line does not.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,77 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
+    // The weight for the document should be updated to double(-1).
     TEST_EQUAL_DOUBLE(i.get_weight(), double(-1));
-    // checking whether the max_attained attribute was updated correctly
+    // max_attained for the weight of the single document {double(-1)} = double(-1).
```
Similarly, kill this comment.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,77 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
+    // max_attained for the weight of the single document {double(-1)} = double(-1).
     TEST_EQUAL_DOUBLE(mymset.get_max_attained(), double(-1));
+    // max_possible should be max(max_possible of the previous weights, max_attained for the new weights).
+    TEST_EQUAL_DOUBLE(mymset.get_max_possible(), max(mymset.get_max_attained(), mymset.get_max_possible()));
```
While this tests that in this case `max_possible` is correct, it's probably the wrong way to test it generally, and doesn't actually ensure that the `max_possible` code is exercised fully. (I don't think the others do either.) For that, you'd need to replace with a weight that's higher than the previous `max_possible`, and then explicitly test equality against that. (You could do that by adding `1` to the previous `max_possible` when constructing the `vector<double>`.)

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,77 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
     Xapian::MSet mymset = enquire.get_mset(0, 10);
 
-    // Checking for exception when the number of weights in the input exceeds the number of documents
+    // Checking for exception when the number of weights in the input exceeds the number of documents.
```
Again, none of these comments, including this one, says anything beyond the code it comments.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,77 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
     }
-    // checking whether the max_attained attribute was updated correctly.
+    // max_attained of the updated weights {double(-2), double(-1)} should be double(-1).
     TEST_EQUAL_DOUBLE(mymset.get_max_attained(), double(-1));
```
`1.0` is a `double` constant in C++ (see this [somewhat dense documentation](http://en.cppreference.com/w/cpp/language/floating_literal)).

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,77 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
-    // checking whether the max_attained attribute was updated correctly.
+    // max_attained of the updated weights {double(-2), double(-1)} should be double(-1).
     TEST_EQUAL_DOUBLE(mymset.get_max_attained(), double(-1));
+    TEST_EQUAL_DOUBLE(mymset.get_max_possible(), max(mymset.get_max_attained(), mymset.get_max_possible()));
```
Again, you should just test the constant directly, which will be the old `max_possible`. You still need a test that it changes when higher weights are replaced in.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,77 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
-    // checking for the correct order of documents and their respective weights.
+    /* checking for the correct order of documents and their respective weights
+     * correct order should be 1.{docids[1], double(-1)} 2.{docids[0],double(-2)}.
+     */
```
You don't need the first line of this, and I'd probably state explicitly that the documents should have been reversed. Note that a less "clever" way of checking, by pulling `mymset.begin()`, checking that with constant values, then `m++` and checking the next one, would probably be clearer than the loop you're doing. (Explicitly unwinding the loop when there are only two passes is a good compromise if it's clearer, which is particularly likely in tests.)

AyushP123 replied:
:+1: ..... I have added the line "The order of documents should have been reversed." after calling sort_by_relevance.

jaylett commented on xapian-core/api/omenquire.cc:
```diff
@@ -467,13 +467,18 @@ MSet::Internal::read_docs() const
+     * While assigning new weights max_attained is updated to reflect the max_attained in the new weights.
+     * While assigning new weights max_possible is updated to reflect the max of max_attained and max_possible of previous weights.
+     * Ideally the max_possible should be the maximum possible weight that can be assigned by the weight assigning algorithm
+     * but since it is not practical to calculate max possible weight for different algorithms we use this approach.
```
This is fairly verbose (for instance the first two lines say basically the same thing), and still misses the key point: this method _must_ be called in turn on every item in the `MSet` to avoid miscalculation of `max_attained`.

It might be clearer if you put the relevant commentary by the code it refers to. Talk about why the `max_attained` updating works (providing this is called appropriately) next to that code. Talk about why we do `max_possible` like this by the code that implements it. The note about correct calling should certainly be in the docstring for this (internal) method, as well as here, since that's where people are more likely to expect that kind of caveat.

Finally, it is _sometimes_ practical to calculate the `max_possible` weight for different re-sorting algorithms, so you should say something like "but since it may not be possible to calculate…".

AyushP123 replied:
I have resolved the issues regarding the tests .... will work on writing a better comments for set_item_wieght

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,46 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
-
+    TEST_EQUAL_DOUBLE(i.get_weight(), new_weight);
+    TEST_EQUAL_DOUBLE(mymset.get_max_attained(), new_weight);
+    TEST_EQUAL_DOUBLE(mymset.get_max_possible(), max(old_max_possible, mymset.get_max_possible()));
```
This is now very odd! You should just check whether the new `max_possible` is `old_max_possible`. What you're doing here actually tests whether `mymset.get_max_possible() >= old_max_possible`, but in a rather convoluted way!

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,46 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
-
-    // assigning new weight of -1 to the document.
-    vector<double> weights {-1};
+    double new_weight = (mymset.begin()).get_weight() - 0.01;
```
I meant actually to find out what that value was, rather than calculating it. It's probably fine here, but it causes you problems in a later test, where you just set all the weights to that first (highest) original weight…

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,46 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
-
-    // Checking for exception when the number of weights in the input exceeds the number of documents
-    vector<double> weights {-1, -2};
+    vector<double> weights {(mymset.begin()).get_weight(), (mymset.begin()).get_weight()};
```
Just use `{ 1.0, 2.0 }` here; it's the wrong length, so any valid `double` values are fine.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,46 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
-    TEST_EQUAL_DOUBLE(mymset.get_max_attained(), double(-1));
+    TEST_EQUAL_DOUBLE((mymset.begin()).get_weight(), new_weight2);
+    TEST_EQUAL_DOUBLE((++mymset.begin()).get_weight(), new_weight1);
+    TEST_EQUAL_DOUBLE(mymset.get_max_attained(), max(new_weight1, new_weight2));
```
RHS will always be `new_weight2`, so use that explicitly.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,46 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
+    TEST_EQUAL_DOUBLE((mymset.begin()).get_weight(), new_weight2);
+    TEST_EQUAL_DOUBLE((++mymset.begin()).get_weight(), new_weight1);
+    TEST_EQUAL_DOUBLE(mymset.get_max_attained(), max(new_weight1, new_weight2));
+    TEST_EQUAL_DOUBLE(mymset.get_max_possible(), max(old_max_possible, mymset.get_max_possible()));
```
Again, without knowing what `old_max_possible` is, you don't know if you're exercising the code correctly. Find out what the actual value will be (by hand) and hard code things higher than it as weights to ensure that it moves. `max()` does not belong in this test!

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -321,79 +321,46 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
-    // checking whether the max_attained attribute was updated correctly.
-    TEST_EQUAL_DOUBLE(mymset.get_max_attained(), double(-1));
+    TEST_EQUAL_DOUBLE((mymset.begin()).get_weight(), new_weight2);
+    TEST_EQUAL_DOUBLE((++mymset.begin()).get_weight(), new_weight1);
```
Declare and use an iterator to make this clearer and less dense. You also need to dereference the iterator at both points to check that the docids are now in the new order.

AyushP123 replied:
:+1:  ..... I have used a for loop to resolve this

AyushP123 commented:
@ojwb , @jaylett .... I really apologize for bugging you again and again with this PR and I know this is not the way a contribution should take place .... I will try in the future for cleaner commits and more focused approach .... I have updated the tests and the comments .... please let me know if any further changes are required .... Also if you two permit I would like to open an alternate PR in which I have implemented in my opinion a better approach .... Although we also have #141 to review and merge

jaylett commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,37 @@ MSet::Internal::read_docs() const
+void
+MSet::Internal::set_item_weight(Xapian::doccount i, double wt_)
+{
+    /* max_attained is updated assuming that set_item_weight is called on every Xapian::doccount from 0 to mset.size() in an ascending order.
```
Surely `mset.size() - 1`, so perhaps just say "on every MSet item from 0 up".

AyushP123 replied:
:+1:

jaylett commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,37 @@ MSet::Internal::read_docs() const
+    /* While assigning new weights max_possible is updated as the maximum of max_attained at Xapian::doccount i
+     * and max_possible at Xapian::doccount i-1. While calculating max_possible for Xapian::doccount 0 the max_possible
+     * of the previous weights is taken into account.
+     * max_possible is set as the maximum of the new weights as well as the old weights.
```
The first four lines of this aren't needed and just confuse matters.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,37 @@ MSet::Internal::read_docs() const
+     * of the previous weights is taken into account.
+     * max_possible is set as the maximum of the new weights as well as the old weights.
+     * Ideally the max_possible should be the maximum possible weight that can be assigned by the resorting algorithm
+     * but since it is not possible to calculate max possible weight for different algorithms we use this approach.
```
I've asked that this be made softer before: it _is possible for some algorithms_, but we aren't even trying to at the moment, so we should say something like "it is not always possible".

AyushP123 replied:
:+1: .... I have updated the docstring and the comments ..... I will use them as they are if you approve

jaylett commented on xapian-core/api/omenquireinternal.h:
```diff
@@ -313,6 +313,21 @@ class MSet::Internal : public Xapian::Internal::intrusive_base {
 	 */
 	void fetch_items(Xapian::doccount first, Xapian::doccount last) const;
+
+	/** Updates the weight corresponding to the document indexed at position Xapian::docount i with double wt_.
```
You don't need the types here, just the variable names would make it clear enough.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/api/omenquireinternal.h:
```diff
@@ -313,6 +313,21 @@ class MSet::Internal : public Xapian::Internal::intrusive_base {
+	 *  max_attained and max_possible.
+	 *
+	 *  @param i Xapian::docount i, position of the document whose weight is being updated.
+	 *  @param wt_ double wt_, new weight assigned to the document.
```
This isn't quite how `@param` works; check out the [relevant doxygen documentation](https://www.stack.nl/~dimitri/doxygen/manual/commands.html#cmdparam) for details.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/include/xapian/mset.h:
```diff
@@ -47,6 +48,9 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
     // Helper function for fetch() methods.
     void fetch_(Xapian::doccount first, Xapian::doccount last) const;
 
+    // Helper function for replace_weights(Iterator first, Iterator last)
```
We should probably say "See the note about correct usage in the docstring for `Xapian::Internal::MSet::set_item_weight()`" or something similar.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -319,3 +319,49 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
+    Xapian::Enquire enquire(db);
+    enquire.set_query(Xapian::Query("word"));
+    Xapian::MSet mymset = enquire.get_mset(0, 10);
+    vector<Xapian::docid> docids {*(mymset.begin()), *(++mymset.begin())};
```
You can use `mymset[0]` and `mymset[1]`, which will likely look a little less strangely than this.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -319,3 +319,49 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
+    mymset.replace_weights(weights.begin(), weights.end());
+    mymset.sort_by_relevance();
+    int k = 1;
+    for (Xapian::MSetIterator m = mymset.begin(); m != mymset.end(); ++m,--k) {
```
Space after `,` in the iteration expression, please.

AyushP123 replied:
:+1:

jaylett commented on xapian-letor/bin/xapian-rank.cc:
```diff
@@ -207,13 +206,16 @@ try {
+    cout << "Docids after re-ranking by LTR model:\n" << endl;
 
+    for (Xapian::MSetIterator i = mset.begin(); i != mset.end(); ++i) {
+
```
This blank line isn't needed.

AyushP123 replied:
:+1:

AyushP123 replied:
:+1:

jaylett commented on xapian-letor/ranker/ranker.cc:
```diff
@@ -351,19 +351,33 @@ Ranker::labelcomparer(const FeatureVector & firstfv, const FeatureVector& second
+    bool operator==(const ScoreIterator& o) { return it == o.it; }
+
+    Xapian::doccount operator-(const ScoreIterator& o) { return Xapian::doccount(it - o.it); }
+
```
This blank line isn't needed.

AyushP123 replied:
:+1:

AyushP123 replied:
:+1:

ojwb commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,35 @@ MSet::Internal::read_docs() const
+    else
+	max_attained = max(max_attained, wt_);
+    /* Ideally the max_possible should be the maximum possible weight that
+     * can be assigned by the resorting algorithm, but since it is not always
```
ITYM "reranking algorithm"

AyushP123 replied:
:+1:

ojwb commented on xapian-core/api/omenquireinternal.h:
```diff
@@ -313,6 +313,21 @@ class MSet::Internal : public Xapian::Internal::intrusive_base {
+	 *  Updates the max_possible and max_attained for the MSet.
+	 *  set_item_weight must be called upon to update the weight of every document in MSet
+	 *  in the same order as that of the documents present in MSet to avoid miscalculation of
+	 *  max_attained and max_possible.
```
There seems to be a word missing after "must be called upon".

I'm not clear why the "same order" part would make any difference here.

AyushP123 replied:
:+1: ....have eliminated "upon" .... will use the updated docstring if you approve

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +80,36 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+	Xapian::doccount i = 0;
+	while (first != last) {
+	    set_item_weight(i, *first);
+	    ++i, ++first;
```
Don't use `,` like this - make it two statements.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +80,36 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+     *
+     *  @exception Xapian::InvalidArgument is thrown if the total number of
+     *  elements in the input dont match the total number of documents in MSet.
+     *
```
"dont" is missing an apostrophe ("don't"), but actually it should be "doesn't" (it's the number doesn't match, so singular).

Extra content-less line at end of comment block.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +80,36 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+
+    /**
+     * Sorts the list of documents in MSet according to their weights. Use after calling MSet::replace_weights.
+     */
```
This doc comment is poorly formatted.

The short description (i.e. the first sentence here) should be on the first line, then there should be a paragraph break between that and the extended description (which would avoid the overly long line).

The naming of `MSet::sort_by_relevance()` seems a bit too similar to `Enquire::set_sort_by_relevance()` - I'd be happy if it was clearer what this method was for from the name.  The obvious alternative is `resort_by_relevance()`, but "resort" is easy to misread as "resort" rather than "re-sort", and `re_sort_by_relevance()` looks weird.  Perhaps `rerank_by_relevance()` is better?  Thoughts?

AyushP123 replied:
@jaylett, @ojwb  .... whichever name the two of you are okay with.

ojwb commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -319,3 +319,49 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
     return true;
 }
+
+DEFINE_TESTCASE(replace_weights, backend) {
```
Conventionally this would be named `replace_weights1`, especially as there's a `replace_weights2`.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -319,3 +319,49 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
+    // old_max_possible = 0.269763689697702
+    double old_max_possible = mymset.get_max_possible();
+    double new_weight = 0.25;
+    vector<double> weights {new_weight};
```
You don't need the extensibility of a vector here, so using `std::array` is better: http://en.cppreference.com/w/cpp/container/array

```
array<double, 1> weights = {{ new_weight }};
```

AyushP123 replied:
:+1:  ..... I am using a static array as we have later discussed.

ojwb commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -319,3 +319,49 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
+    double new_weight = 0.25;
+    vector<double> weights {new_weight};
+    mymset.replace_weights(weights.begin(), weights.end());
+    Xapian::MSetIterator i = mymset.begin();
```
Should probably `TEST(i != mymset.end());` to make sure it's valid to dereference.

AyushP123 replied:
:+1:

ojwb commented on xapian-letor/include/xapian-letor/ranker.h:
```diff
@@ -117,9 +117,9 @@ class XAPIAN_VISIBILITY_DEFAULT Ranker : public Xapian::Internal::intrusive_base
-				    const Xapian::FeatureList & flist = Xapian::FeatureList());
+    void rank(Xapian::MSet & mset,
+	      const std::string & model_key = std::string(),
+	      const Xapian::FeatureList & flist = Xapian::FeatureList());
```
`@return` above is now incorrect.

Please check the diff against the branch point and fix any other instances of this too.

AyushP123 replied:
:+1:

jaylett commented:
That name was my suggestion, by analogy with the Enquire method. I'm not keen on rerank when we talk about sorting elsewhere... But I haven't come up with any name that is neither cumbersome nor potentially ambiguous.On 15 Jun 2017 9:41 p.m., Olly Betts <notifications@github.com> wrote:@ojwb commented on this pull request.



In xapian-core/include/xapian/mset.h:
> +    template <typename Iterator>
+    void replace_weights(Iterator first, Iterator last)
+    {
+	if (last - first != size()) {
+	    throw Xapian::InvalidArgumentError("Number of weights assigned doesn't match the number of items.");
+	}
+	Xapian::doccount i = 0;
+	while (first != last) {
+	    set_item_weight(i, *first);
+	    ++i, ++first;
+	}
+    }
+
+    /**
+     * Sorts the list of documents in MSet according to their weights. Use after calling MSet::replace_weights.
+     */

This doc comment is poorly formatted.
The short description (i.e. the first sentence here) should be on the first line, then there should be a paragraph break between that and the extended description (which would avoid the overly long line).
The naming of MSet::sort_by_relevance() seems a bit too similar to Enquire::set_sort_by_relevance() - I'd be happy if it was clearer what this method was for from the name.  The obvious alternative is resort_by_relevance(), but "resort" is easy to misread as "resort" rather than "re-sort", and re_sort_by_relevance() looks weird.  Perhaps rerank_by_relevance() is better?  Thoughts?

—You are receiving this because you were mentioned.Reply to this email directly, view it on GitHub, or mute the thread.

ojwb commented:
It's true we don't talk about reranking elsewhere in the xapian-core API, but then we don't do reranking elsewhere in it currently.  But indeed reranking vs sorting seems a bit inconsistent.

jaylett commented on xapian-letor/ranker/ranker.cc:
```diff
@@ -351,19 +351,35 @@ Ranker::labelcomparer(const FeatureVector & firstfv, const FeatureVector& second
+Ranker::rank(Xapian::MSet & mset, const string & model_key, const Xapian::FeatureList & flist)
 {
+    if (mset.empty()) {
+	exit(1);
```
No. As repeatedly discussed, this should shortcut the following code and just `return`, because the empty `MSet` is already in its reranked state, since any ranking or sorting of an empty list is the same.

ojwb replied:
You should not print anything from the library - you don't know anything about the program which is using the library.

If it's a GUI app, the output will likely vanish into `~/.xsession-errors` which is so full of junk from Gtk+ these days that basically nobody ever looks at it.  It's where error messages go to quietly die.

If it's a command-line program, then things aren't much better - if it's using curses or similar, just writing to stdout will corrupt the display.  If it's just using the console in a line-based way, you might actually manage to write output to the screen in a way the end user can see.  But the program should decide about that, not the library author.

Or the program may be a daemon and have closed all unwanted file descriptors on start-up.  In that case, fd 1 (which the the fd which `stdout` goes to) may have been reused, so by writing to `stdout` you may dump your output into a file the program is writing to and corrupt it.

Calling `exit(1)` is another thing we shouldn't do.  It causes the program to exit without any chance for vetoing that.  So if the program is some sort of editor and has an unsaved file loaded, you just threw away that unsaved file.

Actual error conditions in the library should throw an exception, not write output to `stdout` or `stderr`, and not call functions like `exit()` and `abort()` which cause the program to exit.  User code can catch and handle the exception in an appropriate way - perhaps showing the user a dialog box, or cleaning up properly before bailing out.  The program knows what's appropriate - we don't.

**HOWEVER**, the `MSet` being empty should not **NOT** an error condition.

It's certainly an edge case, but there's a perfectly valid reranking for an empty `MSet` - an empty reranking.  So as we keep saying, let's just return that rather than turning a case which has an obvious and reasonable behaviour into an error case.  If we make this an error case, then any user code using the xapian-letor API will need to check `mset.empty()` (or equivalent) before calling letor to rerank.  There's no need to impose that requirement - we can just give them back an empty rerank list, they can apply that to get a reranked `MSet` (which will still be empty).

If you were implementing addition, you wouldn't make adding zero to something an error, just because it's trivially doing nothing.  If you did that, then you force everyone to write:

```
if (b != 0) {
    a = a + b;
}
```

Which they'd rightly be annoyed by.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/api/omenquireinternal.h:
```diff
@@ -313,6 +313,21 @@ class MSet::Internal : public Xapian::Internal::intrusive_base {
+
+	/// Sorts the MSet::Internal::items by their corresponding weights.
+	void sort_by_relevance();
+
```
Unwanted line.

AyushP123 replied:
:+1:

jaylett commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +80,38 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
     /// Destructor.
     ~MSet();
 
+    /** Assigns new weights and updates MSet.
```
Line needed after first line of doxygen docstring.

AyushP123 replied:
:+1: have added a line but since I have not changed line 83 it is not showing outdated

jaylett commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +80,38 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+
+    /**
+     * Sorts the list of documents in MSet according to their weights.
+     * Use after calling MSet::replace_weights.
```
Line needed after first line of doxygen docstring.

ojwb replied:
Also the brief description should be on the first line (i.e. `/** Sorts the list of documents in the MSet according to their weights.`)

AyushP123 replied:
:+1:

jaylett commented on xapian-core/api/omenquireinternal.h:
```diff
@@ -313,6 +313,21 @@ class MSet::Internal : public Xapian::Internal::intrusive_base {
 	 */
 	void fetch_items(Xapian::doccount first, Xapian::doccount last) const;
+
+	/** Updates the weight corresponding to the document indexed at position i with wt_.
```
Line needed after first line of doxygen docstring.

AyushP123 replied:
:+1:

ojwb commented on xapian-letor/ranker/ranker.cc:
```diff
@@ -351,19 +351,35 @@ Ranker::labelcomparer(const FeatureVector & firstfv, const FeatureVector& second
+    if (mset.empty()) {
+	return;
+    }
     LOGCALL(API, std::vector<Xapian::docid>, "Ranker::rank", mset | model_key | flist);
```
`LOGCALL` should be the first thing in the function (didn't I already say that?)

Also, you've change the return type, but not updates the `LOGCALL` to match.

AyushP123 replied:
:+1:

ojwb commented on xapian-letor/include/xapian-letor/ranker.h:
```diff
@@ -114,12 +114,12 @@ class XAPIAN_VISIBILITY_DEFAULT Ranker : public Xapian::Internal::intrusive_base
-     *  @exception FileNotFoundError will be thrown if file not found at supplied path
-     *  @return A vector of docids after ranking.
+     *  @exception FileNotFoundError will be thrown if file not found at supplied path.
+     *  @exception InvalidArgumentError will be thrown if the supplied MSet is empty.
```
Drop: `@exception InvalidArgumentError will be thrown if the supplied MSet is empty.`

AyushP123 replied:
:+1:

ojwb commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -319,3 +319,50 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
+    // old_max_possible = 0.269763689697702
+    double old_max_possible = mymset.get_max_possible();
+    double new_weight = 0.125;
+    double weights [1] = {new_weight};
```
There shouldn't be a space before `[` (also below).

AyushP123 replied:
:+1:

ojwb commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,35 @@ MSet::Internal::read_docs() const
+MSet::Internal::set_item_weight(Xapian::doccount i, double wt_)
+{
+    /* max_attained is updated assuming that set_item_weight is called on every MSet item from 0 up.
+     * While assigning new weights max_attained is updated as the maximum of the new weights set till Xapian::doccount i.
```
Comment needs wrapping to fit in 80 columns.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/api/omenquireinternal.h:
```diff
@@ -313,6 +313,22 @@ class MSet::Internal : public Xapian::Internal::intrusive_base {
+	 *  Updates the max_possible and max_attained for the MSet.
+	 *
+	 *  set_item_weight must be called to update the weight of every document in MSet
+	 *  in the same order as that of the documents present in MSet to avoid miscalculation of
```
Line above needs wrapping.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +80,43 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+	if (last - first != size()) {
+	    throw Xapian::InvalidArgumentError("Number of weights assigned"
+					       " doesn't match the number of "
+					       "items.");
```
I'd suggest consistently breaking the string after a space, as that's how we usually do it (currently you have one split before then one after).

Also, we don't usually include a full-stop on an exception message.  It's not entirely consistent currently, but a quick grep suggests it's about 94% without.

AyushP123 replied:
:+1:

ojwb commented on xapian-core/tests/api_sorting.cc:
```diff
@@ -319,3 +319,51 @@ DEFINE_TESTCASE(sortfunctorremote1, remote) {
+    // old_max_possible, max_attained = 0.269763689697702
+    double old_max_possible = mymset.get_max_possible();
+    double new_weight = 0.125;
+    double weights[1] = {new_weight};
```
If you make an array `static` and `const` then it can be initialised at compile time rather than at run time.  In this case you'll also want to make `new_weight` `const`.

Also, it's probably better to let the compiler fill in the number of elements, so:

```
static const weights[] = {new_weight};
```

More instances in the testcases below.

AyushP123 replied:
:+1:

ojwb commented on xapian-letor/include/xapian-letor/ranker.h:
```diff
@@ -37,36 +37,57 @@
+			   Xapian::doccount msetsize,
+			   const std::string & filename = "training-data.txt",
+			   const Xapian::FeatureList & flist =
+			   Xapian::FeatureList());
```
I'd take advantage of us being inside `namespace Xapian {...}` here and drop the `Xapian::` from `Xapian::FeatureList` here so this line doesn't need wrapping.

In a case which really needed wrapping, the default value ought to be indented for clarity.

AyushP123 replied:
:+1:

ojwb commented on xapian-letor/include/xapian-letor/ranker.h:
```diff
@@ -145,30 +191,48 @@ class XAPIAN_VISIBILITY_DEFAULT Ranker : public Xapian::Internal::intrusive_base
 
   protected:
     /// Method to train the model. Overrided in ranker subclass.
```
"Overridden"

AyushP123 replied:
:+1:

ojwb commented on xapian-letor/include/xapian-letor/ranker.h:
```diff
@@ -82,93 +102,136 @@ class XAPIAN_VISIBILITY_DEFAULT Ranker : public Xapian::Internal::intrusive_base
   protected:
-    /// Method to train the model. Overrided in ranker subclass.
-    virtual void train_model(const std::vector<Xapian::FeatureVector> & training_data) = 0;
+    /// Method to train the model. Overriden in ranker subclass.
```
Close, but "overridden" has a double "d".  Also below.

AyushP123 replied:
:+1:

ojwb commented on xapian-letor/include/xapian-letor/ranker.h:
```diff
@@ -37,36 +37,57 @@
+     *				file and database size.
+     *  @param  filename	Filename path where the training file has to
+     *				be stored. Default is "./training-data.txt".
+     *  @param  flist		Xapian::FeatureList object definining what set
```
"defining" - this typo seems to be repeated a number of times below.

AyushP123 replied:
:+1:

AyushP123 commented on xapian-core/api/omenquire.cc:
```diff
@@ -182,6 +182,12 @@ MSet::fetch_(Xapian::doccount first, Xapian::doccount last) const
 }
 
+void
+MSet::set_new_weights_helper(double wt, int i)
```
:+1:

AyushP123 commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,23 @@ MSet::Internal::read_docs() const
+}
+
+void
+MSet::Internal::item_set_weight(Xapian::Internal::MSetItem& item_,double wt_)
```
:+1:

AyushP123 commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,23 @@ MSet::Internal::read_docs() const
+void
+MSet::Internal::re_rank()
+{
+    std::sort(items.begin(),items.end(),[](const Xapian::Internal::MSetItem& x,const Xapian::Internal::MSetItem& y){return x.wt>y.wt;});
```
:+1:

AyushP123 commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +78,31 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+     */
+    template <typename Iterator>
+    void set_new_weights(Iterator first, Iterator last) {
+	int i = 0;
```
:+1:

AyushP123 commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +78,31 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+	int i = 0;
+	if (std::distance(first, last) != size())
+	{
+	    printf("Number of weights assigned don't match the number of items.Aborting");
```
:+1:  .... throwing Xapian::InvalidArgumentError

AyushP123 commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +78,31 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+	for (Iterator it = first;it != last;++it)
+	{
+	    set_new_weights_helper(*it, i);
+	    ++i;
```
:+1: .... have replaced for loop with while loop and eliminated this problem later as suggested by @ojwb

AyushP123 commented on xapian-core/include/xapian/mset.h:
```diff
@@ -47,6 +47,9 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
     void fetch_(Xapian::doccount first, Xapian::doccount last) const;
 
+    // Helper function for set_new_weights(Iterator first,Iterator last)
+    void set_item_weight(int i, double wt);
```
:+1:

AyushP123 commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,28 @@ MSet::Internal::read_docs() const
+void
+MSet::Internal::set_item_weight(int i, double wt_)
+{
+    set_item_weight_(items[i], wt_);
```
:+1:

AyushP123 commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,28 @@ MSet::Internal::read_docs() const
+MSet::Internal::set_item_weight(int i, double wt_)
+{
+    set_item_weight_(items[i], wt_);
+    if(i == 0)
```
:+1:

AyushP123 commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,28 @@ MSet::Internal::read_docs() const
+    set_item_weight_(items[i], wt_);
+    if(i == 0)
+	max_attained = wt_;
+    max_attained = max(max_attained, wt_);
```
:+1:

AyushP123 commented on xapian-core/api/omenquire.cc:
```diff
@@ -452,6 +464,28 @@ MSet::Internal::read_docs() const
+MSet::Internal::sort_by_relevance()
+{
+    std::sort(items.begin(),items.end(),
+	      [](const Xapian::Internal::MSetItem& x, const Xapian::Internal::MSetItem& y){ return x.wt > y.wt; });
```
hopefully I have cleared this .... could you review it and let me know if this has been resolved??

ojwb replied:
Yes, though actually this should fall back to comparing by docid if the weights are equal so that this always provides a stable sort order - e.g. look at the comparisons in `matcher/msetcmp.cc`:

```
// Order by relevance, then docid.
template<bool FORWARD_DID> bool
msetcmp_by_relevance(const Xapian::Internal::MSetItem &a,
                     const Xapian::Internal::MSetItem &b)
{
    if (a.wt > b.wt) return true;
    if (a.wt < b.wt) return false;
    return msetcmp_by_did<FORWARD_DID, true>(a, b);
}
```

It probably actually makes sense to just use `get_mset_cmp_function(Xapian::Enquire::Internal::REL, true, false)` to get a suitable comparator which does that and which we know works consistently with standard `MSet` sorting (the `true` hard wires the forward/reverse ordering of docids - I don't think it matters to respect the order specified when the `MSet` was created here, though we could get it from the `Enquire` object which the `MSet` has a reference to).

AyushP123 replied:
@ojwb do you mean like creating a `MSetCmp mcmp(get_msetcmp_function(Xapian::Enquire::Internal::REL, true, false));
`
and then passing it to `std::sort(items.begin(), items.end(), mcmp);`. I will have to include "msetcmp.h"

AyushP123 replied:
:+1:

jaylett commented on xapian-letor/include/xapian-letor/ranker.h:
```diff
@@ -37,36 +37,56 @@
-    /** Utility function that prepares the 'training-data.txt' file in the current working directory. This file is used to train a model which in turn will be used to
-     *  assign scores to the documents based of Learning-to-Rank model. The file is created in the standard format of Letor training file
-     *  as below:
+    /** Utility function that prepares the 'training-data.txt' file in the
```
Needs a line break somewhere in the start of this docstring.

AyushP123 replied:
:+1:

ojwb commented on xapian-letor/include/xapian-letor/ranker.h:
```diff
@@ -37,36 +37,59 @@
-    /** Utility function that prepares the 'training-data.txt' file in the current working directory. This file is used to train a model which in turn will be used to
-     *  assign scores to the documents based of Learning-to-Rank model. The file is created in the standard format of Letor training file
+    /** Utility function that prepares the 'training-data.txt' file in the
+     *  current working directory.
```
The training data filename is no longer hard-coded, but this needs a tweak to make sense now - something like:

"Utility function that prepares the training data.

The training data file is used to train a model [...]"

ojwb commented:
> @ojwb do you mean like creating a MSetCmp mcmp(get_msetcmp_function(Xapian::Enquire::Internal::REL, true, false)); and then passing it to std::sort(items.begin(), items.end(), mcmp);.

The MSetCmp class really just serves as a way to put the sort function
into a variable, which we don't need to do here - you should just be
able to use get_msetcmp_function(...) as the final parameter to sort,
instead of the current lambda function.

> I will have to include "msetcmp.h"

Yes, of course you will.

AyushP123 commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +80,38 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+     *
+     *  @exception Xapian::InvalidArgument is thrown if the total number of
+     *  elements in the input doesn't match the total number of documents in MSet.
+     *
```
:+1:

AyushP123 commented on xapian-core/include/xapian/mset.h:
```diff
@@ -74,6 +80,43 @@ class XAPIAN_VISIBILITY_DEFAULT MSet {
+     * Sorts the list of documents in MSet according to their weights.
+     *
+     * Use after calling MSet::replace_weights.
+     */
```
:+1:

ojwb commented on xapian-core/api/omenquire.cc:
```diff
@@ -40,6 +40,7 @@
 #include "expand/expandweight.h"
 #include "exp10.h"
 #include "matcher/multimatch.h"
+#include "matcher/msetcmp.h"
```
Again, you've added a header out of sorted order...

ojwb commented:
LGTM - @jaylett are you happy too?

The fixup commits don't autosquash, and there's too daunting a pile to resolve by hand.  I can easily just squash everything down to one commit, though perhaps it is worth rebuilding this as two commits - one for the core API changes (which logically stand by themselves), then one with the changes to letor (which use that new core API).

ojwb commented:
The commit messages don't fully match the conventional formatting.  There should be a blank line after the first line, so that git knows that's a short summary.  As things currently are, git takes the whole first paragraph for the `oneline` format:

```
$ git log --oneline
167bf1e19d0e Letor now updates the original MSet Instead of returning docids sor
ted by the new weights, letor now assigns new weights to the MSet and then re-so
rts it.
8092a04bc17a Add methods to re rank and replace weights in a MSet Two new method
s, replace_weights and sort_by_relevance, have been added to MSet which can be u
sed to replace weights of all the documents in a MSet and then sort the MSet on 
the basis of these new weights.
18c24545de88 Add missing .gitignore entries
a255e1663caf Correct ){ to ) { in a few places
0f8619c76d3b Create constants for glass block size min and max
119f1ce8d50f [CI] Add check for generated files not in .gitignore
1b2e6891a9b1 Ignore new generated file
4848901bd2d6 Add clustering API
3d07e0f8b00b xapian-letor refactoring
f81ab8658613 Fix build error in previous commit
[...]
```

A more minor point is that `Add methods to re rank and replace weights in a MSet` is a couple of characters longer than the ideal 50 characters.  It isn't a big deal if a few extra characters are needed to write something clear rather than cryptic, though here `Add MSet methods to re rank and replace weights` would be clear and fit in the ideal length.

It's easy to adjust commit messages - you can use `git rebase -i HEAD~2` to do an interactive rebase of the last two commits on the current branch, which will open an editor with a list of commits.  Change the word `pick` before the listed commits to `r` (or `reword` - either works) and then save the file and exit the editor.

This will load up the commit messages one-by-one for you to edit.  For each in turn, just make the tweaks and save.  This will rebase the new versions of the commits onto the same place they were before (i.e. the grandparent of current HEAD).  Note that this will cause the commits to get new hashes (the hash includes the parent commit hash, the changes to files in the diff, and metadata for the commit such as the message, author, committer and dates, and you've explicitly changed the message and implicitly changed the commit date) - this means you'll have to force push to update the branch (and so the usually caveats for not force pushing without consent apply).
